### ISSUE TITLE
Bug: [no-unnecessary-type-assertion] false negative with `exactOptionalPropertyTypes` compiler option

### ISSUE THREAD

**Issue by DetachHead** [typescript-eslint/typescript-eslint#5344]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.7.2&sourceType=module&code=CYUwxgNghgTiAEYD2A7AzgF3gMyUgXPAN4CwAUPPFAPyGYwCWKA5uQL7nLpYBGs8AXhx4AdFCpp49JswA+AVxShsTEMHJA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RI4MAF8QOoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3QAacDUjoAHgENmuAsXmoACpliIseAJ4AVPTuFhR6KQF8QloA

### Repro Code

```typescript
declare const foo: {
  a?: string
}
const bar = foo.a as string | undefined
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-unnecessary-type-assertion": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true,
    "exactOptionalPropertyTypes": true
  }
}
```

### Expected Result

`no-unnecessary-type-assertion` error

### Actual Result

no error

### Additional Info

even thought `exactOptionalPropertyTypes` changes the meaning of `a?: string` to be different to `a?: string | undefined`, that difference is not relevant when checking the type of `foo.a`, as it will be `string | undefined` in both cases

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `5.30.6` |
| `@typescript-eslint/parser`            | `5.30.6` |
| `TypeScript`                           | `4,7,2` |
| `ESLint`                               | `8.15.0` |
